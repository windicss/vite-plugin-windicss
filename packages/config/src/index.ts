import fs from 'node:fs'
import path from 'node:path'
import _debug from 'debug'
import type { JITI } from 'jiti'
import _jiti from 'jiti'
import type { FullConfig as WindiCssOptions } from 'windicss/types/interfaces'

let jiti: JITI

export type { WindiCssOptions }

export const defaultConfigureFiles = [
  'windi.config.ts',
  'windi.config.mts',
  'windi.config.js',
  'windi.config.mjs',
  'windi.config.cjs',

  'windicss.config.ts',
  'windicss.config.mts',
  'windicss.config.js',
  'windicss.config.mjs',
  'windicss.config.cjs',

  'tailwind.config.ts',
  'tailwind.config.mts',
  'tailwind.config.js',
  'tailwind.config.mjs',
  'tailwind.config.cjs',
]

export interface LoadConfigurationOptions {
  /**
   * Name for debug
   *
   * @default 'windicss-config'
   * @internal
   */
  name?: string
  /**
   * Config object or path
   */
  config?: WindiCssOptions | string
  /**
   * CWD
   *
   * @default process.cwd
   * @internal
   */
  root?: string
  /**
   * A list of filename of paths to search of config files
   */
  configFiles?: string[]
  /**
   * On loading configuration error
   *
   * @default [throw error]
   */
  onConfigurationError?: (error: unknown) => void
  /**
   * On configure file not found
   *
   * @default [emit warning]
   */
  onConfigurationNotFound?: (filepath: string) => void
}

export function loadConfiguration(options: LoadConfigurationOptions): {
  config: WindiCssOptions
  filepath?: string
  error?: unknown
} {
  if (!jiti)
    jiti = _jiti(process.cwd(), { requireCache: false, cache: false, v8cache: false })

  let resolved: WindiCssOptions = {}
  let configFilePath: string | undefined
  let error: unknown

  const {
    name = 'windicss-config',
    config,
    root = process.cwd(),
    configFiles: configureFiles = defaultConfigureFiles,
    onConfigurationError = (e) => {
      throw e
    },
    onConfigurationNotFound = (path) => {
      console.warn(`[${name}] config file "${path}" not found, ignored`)
    },
  } = options

  const debugConfig = _debug(`${name}:config`)

  if (typeof config === 'string' || !config) {
    if (!config) {
      for (const name of configureFiles) {
        const tryPath = path.resolve(root, name)
        if (fs.existsSync(tryPath)) {
          configFilePath = tryPath
          break
        }
      }
    }
    else {
      configFilePath = path.resolve(root, config)
      if (!fs.existsSync(configFilePath)) {
        onConfigurationNotFound(config)
        configFilePath = undefined
      }
    }

    if (configFilePath) {
      try {
        debugConfig('loading from ', configFilePath)

        resolved = jiti(configFilePath)
        if (resolved.default)
          resolved = resolved.default
      }
      catch (e) {
        error = e
        configFilePath = undefined
        resolved = {}
        onConfigurationError?.(e)
      }
    }
  }
  else {
    resolved = config
  }

  debugConfig(resolved)

  return {
    error,
    config: resolved,
    filepath: configFilePath,
  }
}
