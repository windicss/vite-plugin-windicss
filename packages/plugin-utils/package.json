{
  "name": "@windicss/plugin-utils",
  "version": "1.8.6",
  "description": "Common utils for building integrations of Windi CSS",
  "author": "antfu <anthonyfu117@hotmail.com>",
  "license": "MIT",
  "funding": "https://github.com/sponsors/antfu",
  "homepage": "https://github.com/antfu/vite-plugin-windicss",
  "repository": {
    "type": "git",
    "url": "https://github.com/antfu/vite-plugin-windicss"
  },
  "bugs": "https://github.com/antfu/vite-plugin-windicss/issues",
  "keywords": [
    "windicss",
    "tailwindcss"
  ],
  "exports": {
    ".": {
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./*": "./*"
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup src/index.ts --dts --format cjs,esm --no-splitting --external sucrase/register/ts,pug,windicss/utils/style,windicss/utils/parser,windicss/utils",
    "dev": "npm run build -- --watch",
    "prepublishOnly": "npm run build"
  },
  "dependencies": {
    "@antfu/utils": "^0.5.2",
    "@windicss/config": "workspace:*",
    "debug": "^4.3.4",
    "fast-glob": "^3.2.11",
    "magic-string": "^0.26.2",
    "micromatch": "^4.0.5",
    "windicss": "^3.5.6"
  },
  "devDependencies": {
    "pug": "^3.0.2"
  }
}
