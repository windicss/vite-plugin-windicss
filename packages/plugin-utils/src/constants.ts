export const defaultAlias: Record<string, TagNames> = {
  'router-link': 'a',
}

export const preflightTags = ['html', 'body', 'div']

export const htmlTags = [
  'html',
  'body',
  'div',
  'a',
  'abbr',
  'address',
  'area',
  'article',
  'aside',
  'audio',
  'base',
  'basefont',
  'bdo',
  'blink',
  'blockquote',
  'br',
  'button',
  'canvas',
  'caption',
  'center',
  'col',
  'colgroup',
  'command',
  'comment',
  'datalist',
  'dd',
  'del',
  'details',
  'dir',
  'dl',
  'dt',
  'embed',
  'fieldset',
  'figure',
  'b',
  'big',
  'i',
  'small',
  'tt',
  'font',
  'footer',
  'form',
  'frame',
  'frameset',
  'head',
  'header',
  'hgroup',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'hr',
  'isindex',
  'iframe',
  'ilayer',
  'img',
  'input',
  'ins',
  'keygen',
  'keygen',
  'label',
  'layer',
  'legend',
  'li',
  'link',
  'map',
  'mark',
  'marquee',
  'menu',
  'meta',
  'meter',
  'multicol',
  'nav',
  'nobr',
  'noembed',
  'noframes',
  'noscript',
  'object',
  'ol',
  'optgroup',
  'option',
  'output',
  'p',
  'param',
  'cite',
  'code',
  'dfn',
  'em',
  'kbd',
  'samp',
  'strong',
  'var',
  'plaintext',
  'pre',
  'progress',
  'q',
  'ruby',
  'script',
  'section',
  'select',
  'spacer',
  'span',
  's',
  'strike',
  'style',
  'sub',
  'sup',
  'svg',
  'table',
  'tbody',
  'td',
  'textarea',
  'tfoot',
  'th',
  'thead',
  'time',
  'title',
  'tr',
  'u',
  'ul',
  'video',
  'wbr',
  'wbr',
  'xmp',
] as const

export type TagNames = (typeof htmlTags)[number]
