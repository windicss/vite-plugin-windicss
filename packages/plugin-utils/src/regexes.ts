export const regexQuotedString = /(["'`])((?:\\\1|\\\\|\n|\r|(?!<|>|\s\*\/).)*?)\1/mg
export const regexHtmlTag = /<(\w[\w-]*)([\S\s]*?)\/?[^=]>/mg
export const regexClassSplitter = /[\s'"`{}]/g
export const regexClassGroup = /([!\w+-<@][\w+:_/-]*?\w):\(((?:[!\w\s:/\\,%#.$-]|\[.*?\])*?)\)/gm
export const regexAttributifyItem = /(?:\s|^)([\w+:_/-]+)\s*=\s*(['"{])((?:\\\2|\\\\|\n|\r|.)*?)(?:\2|\})/gm

export const regexClassCheck1 = /^!?[a-z\d@<>.+-](?:\([\w,.%#\(\)+-]*\)|[\w:/\\,%#\[\].$-])*$/
export const regexClassCheck2 = /[a-z].*[\w)\]]$/
export const regexClassChecks = [
  regexClassCheck1,
  regexClassCheck2,
]

export const regexSvelteClass = /class:([\w!:/\\,%#\[\].$-]+?)=["']?\{/g

export function validClassName(i: string) {
  return regexClassChecks.every(r => i.length > 2 && i.match(r))
}
