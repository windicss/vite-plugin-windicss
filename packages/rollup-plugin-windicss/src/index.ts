import { Plugin } from 'rollup'
import _debug from 'debug'
import { createUtils, UserOptions } from '@windicss/plugin-utils'
import { createVirtualModuleLoader } from '../../shared/virtual-module'

const NAME = 'rollup-plugin-windicss'

const debug = {
  hmr: _debug(`${NAME}:hmr`),
  css: _debug(`${NAME}:transform:css`),
  group: _debug(`${NAME}:transform:group`),
  alias: _debug(`${NAME}:transform:alias`),
}

function WindiCssRollupPlugin(userOptions: UserOptions = {}): Plugin[] {
  const utils = createUtils(
    userOptions,
    {
      name: NAME,
      onConfigurationError(e) {
        throw e
      },
    },
  )
  const plugins: Plugin[] = []

  plugins.push({
    name: `${NAME}:alias`,
    async transform(code, id) {
      await utils.ensureInit()
      if (!utils.isDetectTarget(id))
        return
      debug.alias(id)
      return utils.transformAlias(code, true)
    },
  })

  if (userOptions.transformGroups !== false) {
    plugins.push({
      name: `${NAME}:groups`,
      async transform(code, id) {
        await utils.ensureInit()
        if (!utils.isDetectTarget(id))
          return
        debug.group(id)
        return utils.transformGroups(code, true)
      },
    })
  }

  plugins.push({
    name: `${NAME}:entry`,

    ...createVirtualModuleLoader({ utils }),
  })

  if (userOptions.transformCSS !== false) {
    plugins.push({
      name: `${NAME}:css`,
      async transform(code, id) {
        await utils.ensureInit()
        if (!utils.isCssTransformTarget(id))
          return
        debug.css(id)
        return utils.transformCSS(code, id)
      },
    })
  }

  return plugins
}

export * from '@windicss/plugin-utils'
export default WindiCssRollupPlugin
