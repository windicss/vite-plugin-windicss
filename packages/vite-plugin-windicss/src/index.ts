import { resolve } from 'node:path'
import type { Plugin, ResolvedConfig, ViteDevServer } from 'vite'
import _debug, { log } from 'debug'
import type { UserOptions, WindiPluginUtils, WindiPluginUtilsOptions } from '@windicss/plugin-utils'
import { createUtils } from '@windicss/plugin-utils'
import { MODULE_ID_VIRTUAL_PREFIX, createVirtualModuleLoader } from '../../shared/virtual-module'
import { createDevtoolsPlugin } from './devtools'
import { NAME } from './constants'
import { getCssModules, reloadChangedCssModules } from './modules'

const debug = {
  hmr: _debug(`${NAME}:hmr`),
  css: _debug(`${NAME}:transform:css`),
  group: _debug(`${NAME}:transform:group`),
  alias: _debug(`${NAME}:transform:alias`),
  memory: _debug(`${NAME}:memory`),
}

function VitePluginWindicss(userOptions: UserOptions = {}, utilsOptions: WindiPluginUtilsOptions = {}): Plugin[] {
  let utils: WindiPluginUtils
  // let viteConfig: ResolvedConfig
  let server: ViteDevServer | undefined
  let viteConfig: ResolvedConfig

  const plugins: Plugin[] = []

  // transform alias
  plugins.push({
    name: `${NAME}:alias`,
    enforce: 'pre',
    configResolved(_config) {
      viteConfig = _config
    },
    async transform(code, id) {
      await utils.ensureInit()
      if (!utils.isDetectTarget(id))
        return
      debug.alias(id)
      return utils.transformAlias(code, !!viteConfig.build.sourcemap)
    },
  })

  // Utilities grouping transform
  if (userOptions.transformGroups !== false) {
    plugins.push({
      name: `${NAME}:groups`,
      enforce: 'pre',
      async transform(code, id) {
        await utils.ensureInit()
        if (!utils.isDetectTarget(id))
          return
        debug.group(id)
        return utils.transformGroups(code, !!viteConfig.build.sourcemap)
      },
    })
  }

  // exposing api
  plugins.push({
    name: NAME,
    get api() {
      return utils
    },
  })

  // CSS Entry via virtual module
  plugins.push({
    name: `${NAME}:entry`,
    enforce: 'post',

    configureServer(_server) {
      server = _server
    },

    async configResolved(_config) {
      // viteConfig = _config
      utils = utilsOptions.utils ?? createUtils(userOptions, {
        name: NAME,
        root: _config.root,
        onConfigurationError(e) {
          if (_config.command === 'build') {
            throw e
          }
          else {
            console.error(`[${NAME}] Error on loading configurations`)
            console.error(e)
          }
        },
        ...utilsOptions,
      })
      await utils.ensureInit()
    },

    ...createVirtualModuleLoader({
      get utils() { return utils },
      get inHmr() {
        return Boolean(server)
      },
    }) as Omit<Plugin, 'name'>,
  })

  let _cssReloadTask: any
  function reloadCssModules(server: ViteDevServer) {
    clearTimeout(_cssReloadTask)
    _cssReloadTask = setTimeout(() => {
      reloadChangedCssModules(server, utils)
    }, 1)
  }

  // HMR
  plugins.push({
    name: `${NAME}:hmr`,
    apply: 'serve',
    enforce: 'pre',

    async configureServer(_server) {
      server = _server

      await utils.ensureInit()
      if (utils.configFilePath)
        server.watcher.add(utils.configFilePath)

      // NOTE: Track changes to the files so that they are re-scanned as needed.
      // Added files are only detected if the user explicitly enables globbing.
      const supportsGlobs = server.config.server.watch?.disableGlobbing === false
      server.watcher.add(supportsGlobs ? utils.globs : await utils.getFiles())
    },

    async handleHotUpdate({ server, file, read }) {
      // resolve normalized file path to system path
      if (resolve(file) === utils.configFilePath) {
        debug.hmr(`config file changed: ${file}`)
        await utils.init()
        setTimeout(() => {
          log('configure file changed, reloading')
          server.ws.send({ type: 'full-reload' })
        }, 0)
        return getCssModules(server)
      }

      if (!utils.isDetectTarget(file))
        return

      utils.extractFile(await read(), file, true)
        .then((changed) => {
          if (changed) {
            debug.hmr(`refreshed by ${file}`)
            reloadCssModules(server)
          }
        })
    },
  })

  const { transformCSS: transformCSSOptions = true } = userOptions

  const transformCSS = (code: string, id: string) => utils.transformCSS(code, id, {
    onLayerUpdated() {
      if (server)
        reloadCssModules(server)
    },
  })

  // CSS transform
  if (transformCSSOptions === true) {
    plugins.push({
      name: `${NAME}:css`,
      async transform(code, id) {
        await utils.ensureInit()
        if (!utils.isCssTransformTarget(id) || id.startsWith(MODULE_ID_VIRTUAL_PREFIX))
          return
        debug.css(id)
        code = transformCSS(code, id)
        if (viteConfig.build.sourcemap) {
          return {
            code: transformCSS(code, id),
            map: { mappings: '' },
          }
        }
        else {
          return code
        }
      },
    })
  }
  else if (typeof transformCSSOptions === 'string') {
    plugins.push({
      name: `${NAME}:css`,
      enforce: transformCSSOptions,
      transform(code, id) {
        if (!utils.isCssTransformTarget(id) || id.startsWith(MODULE_ID_VIRTUAL_PREFIX))
          return
        debug.css(id, transformCSSOptions)
        code = transformCSS(code, id)
        if (viteConfig.build.sourcemap) {
          return {
            code: transformCSS(code, id),
            map: { mappings: '' },
          }
        }
        else {
          return code
        }
      },
    })
  }

  plugins.push({
    name: `${NAME}:css:svelte`,
    api: {
      sveltePreprocess: {
        style({ content, id }: { content: string; id: string }) {
          return {
            code: transformCSS(content, id),
          }
        },
      },
    },
  })

  plugins.push(...createDevtoolsPlugin({
    get utils() { return utils },
  }))

  return plugins
}

export * from '@windicss/plugin-utils'
export default VitePluginWindicss
