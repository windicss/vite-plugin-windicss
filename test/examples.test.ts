import { resolve } from 'path'
import { describe, expect, it } from 'vitest'
import { createUtils } from '../packages/plugin-utils/src'

describe('example', () => {
  it('should work', async () => {
    const utils = createUtils({}, {
      root: resolve(__dirname, '../examples/vue'),
      onConfigurationError(e) {
        throw e
      },
    })
    await utils.init()
    expect(await utils.generateCSS()).toMatchSnapshot('generated-css')

    // layers
    expect(await utils.generateCSS('base')).toMatchSnapshot('generated-css-base')
    expect(await utils.generateCSS('components')).toMatchSnapshot('generated-css-components')
    expect(await utils.generateCSS('utilities')).toMatchSnapshot('generated-css-utilities')

    expect(utils.classesGenerated).toMatchSnapshot('classes')
    expect(utils.tagsGenerated).toMatchSnapshot('tags')
  })

  it('should allow access to generated tags and classes by a callback', async () => {
    let classes, tags, classesPending, tagsPending
    const utils = createUtils({
      onBeforeGenerate(ctx) {
        classesPending = new Set(ctx.classesPending)
        tagsPending = new Set(ctx.tagsPending)
      },
      onGenerated(ctx) {
        classes = ctx.classes
        tags = ctx.tags
      },
    }, {
      root: resolve(__dirname, '../examples/vue'),
      onConfigurationError(e) {
        throw e
      },
    })
    await utils.init()
    await utils.generateCSS()
    expect(classesPending).toMatchSnapshot('classesPending')
    expect(tagsPending).toMatchSnapshot('tagsPending')
    expect(classes).toMatchSnapshot('classes')
    expect(tags).toMatchSnapshot('tags')
  })

  it('v3', async () => {
    const utils = createUtils(
      {
        preflight: false,
      },
      {
        root: resolve(__dirname, '../examples/windicss-v3'),
        onConfigurationError(e) {
          throw e
        },
      },

    )
    await utils.init()
    const base = await utils.generateCSS('base')
    const components = await utils.generateCSS('components')
    const utilities = await utils.generateCSS('utilities')

    expect(base).toMatchSnapshot('base')
    expect(components).toMatchSnapshot('components')
    expect(utilities).toMatchSnapshot('utilities')
  })
})
