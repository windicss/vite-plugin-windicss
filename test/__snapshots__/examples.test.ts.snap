// Vitest Snapshot v1

exports[`example > should allow access to generated tags and classes by a callback > classes 1`] = `
Set {
  "select-none",
  "p-5",
  "px-5",
  "p-1",
  "text-white",
  "hover:text-green-100",
  "hover:rounded-full",
  "bg-teal-100",
  "hover:w-1/3",
  "btn",
  "bg-black",
  "bg-blue-200",
  "bg-red-200",
  "!bg-blue-200",
  "p-4",
  "pl-100px",
  "bg-yellow-200",
  "mt-2",
}
`;

exports[`example > should allow access to generated tags and classes by a callback > classesPending 1`] = `
Set {
  "select-none",
  "<html",
  "<meta",
  "viewport",
  "<div",
  "app",
  "p-5",
  "<script",
  "module",
  "setup",
  "import",
  "from",
  "./Foo.vue",
  "with",
  "px-5",
  "p-1",
  "text-white",
  "hover:text-green-100",
  "hover:rounded-full",
  "bg-teal-100",
  "hover:w-1/3",
  "<h2",
  "btn",
  "bar",
  "<Foo",
  "<router-link",
  "bg-black",
  "bg-blue-200",
  "bg-red-200",
  "!bg-blue-200",
  "<style",
  "postcss",
  ".bar",
  "@apply",
  "p-4",
  "@screen",
  "scss",
  "stylus",
  "pl-100px",
  "bg-yellow-200",
  "mt-2",
  "createApp",
  "vue",
  "./App.vue",
  "virtual:windi-base.css",
  "virtual:windi-components.css",
  "virtual:windi-utilities.css",
  "virtual:windi-devtools",
  "./main.css",
  "../excluded/included.css",
  "join",
  "path",
  "defineConfig",
  "vite",
  "@vitejs/plugin-vue",
  "vite-plugin-windicss",
  "vite-plugin-restart",
  "export",
  "default",
  "excluded",
  "included.css",
}
`;

exports[`example > should allow access to generated tags and classes by a callback > tags 1`] = `
Set {
  "html",
  "body",
  "div",
  "head",
  "meta",
  "title",
  "script",
  "nav",
  "h2",
  "a",
  "style",
}
`;

exports[`example > should allow access to generated tags and classes by a callback > tagsPending 1`] = `
Set {
  "html",
  "body",
  "div",
  "head",
  "meta",
  "title",
  "script",
  "nav",
  "h2",
  "a",
  "style",
}
`;

exports[`example > should work > classes 1`] = `
Set {
  "select-none",
  "p-5",
  "px-5",
  "p-1",
  "text-white",
  "hover:text-green-100",
  "hover:rounded-full",
  "bg-teal-100",
  "hover:w-1/3",
  "btn",
  "bg-black",
  "bg-blue-200",
  "bg-red-200",
  "!bg-blue-200",
  "p-4",
  "pl-100px",
  "bg-yellow-200",
  "mt-2",
}
`;

exports[`example > should work > generated-css 1`] = `
"/* windicss layer base */
*, ::before, ::after {
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
  border-width: 0;
  border-style: solid;
  border-color: #e5e7eb;
}
* {
  --tw-ring-inset: var(--tw-empty,/*!*/ /*!*/);
  --tw-ring-offset-width: 0px;
  --tw-ring-offset-color: #fff;
  --tw-ring-color: rgba(59, 130, 246, 0.5);
  --tw-ring-offset-shadow: 0 0 #0000;
  --tw-ring-shadow: 0 0 #0000;
  --tw-shadow: 0 0 #0000;
}
:root {
  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;
}
:-moz-focusring {
  outline: 1px dotted ButtonText;
}
:-moz-ui-invalid {
  box-shadow: none;
}
::moz-focus-inner {
  border-style: none;
  padding: 0;
}
::-webkit-inner-spin-button, ::-webkit-outer-spin-button {
  height: auto;
}
::-webkit-search-decoration {
  -webkit-appearance: none;
}
::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}
[type='search'] {
  -webkit-appearance: textfield;
  outline-offset: -2px;
}
abbr[title] {
  -webkit-text-decoration: underline dotted;
  text-decoration: underline dotted;
}
a {
  color: inherit;
  text-decoration: inherit;
}
body {
  margin: 0;
  font-family: inherit;
  line-height: inherit;
}
html {
  -webkit-text-size-adjust: 100%;
  font-family: ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,\\"Segoe UI\\",Roboto,\\"Helvetica Neue\\",Arial,\\"Noto Sans\\",sans-serif,\\"Apple Color Emoji\\",\\"Segoe UI Emoji\\",\\"Segoe UI Symbol\\",\\"Noto Color Emoji\\";
  line-height: 1.5;
}
h2 {
  margin: 0;
  font-size: inherit;
  font-weight: inherit;
}
/* windicss layer components */
.btn {
  --tw-border-opacity: 1;
  border-color: rgba(209, 213, 219, var(--tw-border-opacity));
  border-radius: 0.25rem;
  border-width: 1px;
  display: inline-block;
  margin: 0.5rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  --tw-text-opacity: 1;
  color: rgba(75, 85, 99, var(--tw-text-opacity));
}
.btn:hover {
  --tw-shadow: 0 1px 3px 0 rgb(0 0 0/0.1),0 1px 2px -1px rgb(0 0 0/0.1);
  --tw-shadow-colored: 0 1px 3px 0 var(--tw-shadow-color),0 1px 2px -1px var(--tw-shadow-color);
  -webkit-box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
  box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
}
/* windicss layer utilities */
.bg-teal-100 {
  --tw-bg-opacity: 1;
  background-color: rgba(0, 153, 102, var(--tw-bg-opacity));
}
.bg-black {
  --tw-bg-opacity: 1;
  background-color: rgba(0, 0, 0, var(--tw-bg-opacity));
}
.bg-blue-200 {
  --tw-bg-opacity: 1;
  background-color: rgba(191, 219, 254, var(--tw-bg-opacity));
}
.bg-red-200 {
  --tw-bg-opacity: 1;
  background-color: rgba(254, 202, 202, var(--tw-bg-opacity));
}
.\\\\!bg-blue-200 {
  --tw-bg-opacity: 1 !important;
  background-color: rgba(191, 219, 254, var(--tw-bg-opacity)) !important;
}
.bg-yellow-200 {
  --tw-bg-opacity: 1;
  background-color: rgba(253, 230, 138, var(--tw-bg-opacity));
}
.hover\\\\:rounded-full:hover {
  border-radius: 9999px;
}
.mt-2 {
  margin-top: 0.5rem;
}
.p-5 {
  padding: 1.25rem;
}
.p-1 {
  padding: 0.25rem;
}
.p-4 {
  padding: 1rem;
}
.px-5 {
  padding-left: 1.25rem;
  padding-right: 1.25rem;
}
.pl-100px {
  padding-left: 100px;
}
.text-white {
  --tw-text-opacity: 1;
  color: rgba(255, 255, 255, var(--tw-text-opacity));
}
.hover\\\\:text-green-100:hover {
  --tw-text-opacity: 1;
  color: rgba(209, 250, 229, var(--tw-text-opacity));
}
.select-none {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}
.hover\\\\:w-1\\\\/3:hover {
  width: 33.333333%;
}"
`;

exports[`example > should work > generated-css-base 1`] = `
"/* windicss layer base */
*, ::before, ::after {
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
  border-width: 0;
  border-style: solid;
  border-color: #e5e7eb;
}
* {
  --tw-ring-inset: var(--tw-empty,/*!*/ /*!*/);
  --tw-ring-offset-width: 0px;
  --tw-ring-offset-color: #fff;
  --tw-ring-color: rgba(59, 130, 246, 0.5);
  --tw-ring-offset-shadow: 0 0 #0000;
  --tw-ring-shadow: 0 0 #0000;
  --tw-shadow: 0 0 #0000;
}
:root {
  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;
}
:-moz-focusring {
  outline: 1px dotted ButtonText;
}
:-moz-ui-invalid {
  box-shadow: none;
}
::moz-focus-inner {
  border-style: none;
  padding: 0;
}
::-webkit-inner-spin-button, ::-webkit-outer-spin-button {
  height: auto;
}
::-webkit-search-decoration {
  -webkit-appearance: none;
}
::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}
[type='search'] {
  -webkit-appearance: textfield;
  outline-offset: -2px;
}
abbr[title] {
  -webkit-text-decoration: underline dotted;
  text-decoration: underline dotted;
}
a {
  color: inherit;
  text-decoration: inherit;
}
body {
  margin: 0;
  font-family: inherit;
  line-height: inherit;
}
html {
  -webkit-text-size-adjust: 100%;
  font-family: ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,\\"Segoe UI\\",Roboto,\\"Helvetica Neue\\",Arial,\\"Noto Sans\\",sans-serif,\\"Apple Color Emoji\\",\\"Segoe UI Emoji\\",\\"Segoe UI Symbol\\",\\"Noto Color Emoji\\";
  line-height: 1.5;
}
h2 {
  margin: 0;
  font-size: inherit;
  font-weight: inherit;
}"
`;

exports[`example > should work > generated-css-components 1`] = `
"/* windicss layer components */
.btn {
  --tw-border-opacity: 1;
  border-color: rgba(209, 213, 219, var(--tw-border-opacity));
  border-radius: 0.25rem;
  border-width: 1px;
  display: inline-block;
  margin: 0.5rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  --tw-text-opacity: 1;
  color: rgba(75, 85, 99, var(--tw-text-opacity));
}
.btn:hover {
  --tw-shadow: 0 1px 3px 0 rgb(0 0 0/0.1),0 1px 2px -1px rgb(0 0 0/0.1);
  --tw-shadow-colored: 0 1px 3px 0 var(--tw-shadow-color),0 1px 2px -1px var(--tw-shadow-color);
  -webkit-box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
  box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
}"
`;

exports[`example > should work > generated-css-utilities 1`] = `
"/* windicss layer utilities */
.bg-teal-100 {
  --tw-bg-opacity: 1;
  background-color: rgba(0, 153, 102, var(--tw-bg-opacity));
}
.bg-black {
  --tw-bg-opacity: 1;
  background-color: rgba(0, 0, 0, var(--tw-bg-opacity));
}
.bg-blue-200 {
  --tw-bg-opacity: 1;
  background-color: rgba(191, 219, 254, var(--tw-bg-opacity));
}
.bg-red-200 {
  --tw-bg-opacity: 1;
  background-color: rgba(254, 202, 202, var(--tw-bg-opacity));
}
.\\\\!bg-blue-200 {
  --tw-bg-opacity: 1 !important;
  background-color: rgba(191, 219, 254, var(--tw-bg-opacity)) !important;
}
.bg-yellow-200 {
  --tw-bg-opacity: 1;
  background-color: rgba(253, 230, 138, var(--tw-bg-opacity));
}
.hover\\\\:rounded-full:hover {
  border-radius: 9999px;
}
.mt-2 {
  margin-top: 0.5rem;
}
.p-5 {
  padding: 1.25rem;
}
.p-1 {
  padding: 0.25rem;
}
.p-4 {
  padding: 1rem;
}
.px-5 {
  padding-left: 1.25rem;
  padding-right: 1.25rem;
}
.pl-100px {
  padding-left: 100px;
}
.text-white {
  --tw-text-opacity: 1;
  color: rgba(255, 255, 255, var(--tw-text-opacity));
}
.hover\\\\:text-green-100:hover {
  --tw-text-opacity: 1;
  color: rgba(209, 250, 229, var(--tw-text-opacity));
}
.select-none {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}
.hover\\\\:w-1\\\\/3:hover {
  width: 33.333333%;
}"
`;

exports[`example > should work > tags 1`] = `
Set {
  "html",
  "body",
  "div",
  "head",
  "meta",
  "title",
  "script",
  "nav",
  "h2",
  "a",
  "style",
}
`;

exports[`example > v3 > base 1`] = `
"/* windicss layer base */
"
`;

exports[`example > v3 > components 1`] = `
"/* windicss layer components */
"
`;

exports[`example > v3 > utilities 1`] = `
"/* windicss layer utilities */
[hover\\\\:bg~=\\"gray-400\\"]:hover {
  --tw-bg-opacity: 1;
  background-color: rgba(156, 163, 175, var(--tw-bg-opacity));
}
[hover\\\\:bg~=\\"opacity-10\\"]:hover {
  --tw-bg-opacity: 0.1;
}
.border-gray-200 {
  --tw-border-opacity: 1;
  border-color: rgba(229, 231, 235, var(--tw-border-opacity));
}
.dark .dark\\\\:border-dark-300 {
  --tw-border-opacity: 1;
  border-color: rgba(45, 45, 45, var(--tw-border-opacity));
}
[rounded~=\\"md\\"] {
  border-radius: 0.375rem;
}
.border {
  border-width: 1px;
}
[border~=\\"2\\"] {
  border-width: 2px;
}
.grid {
  display: -ms-grid;
  display: grid;
}
[font~=\\"mono\\"] {
  font-family: ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,\\"Liberation Mono\\",\\"Courier New\\",monospace;
}
[font~=\\"bold\\"] {
  font-weight: 700;
}
[text~=\\"xl\\"] {
  font-size: 1.25rem;
  line-height: 1.75rem;
}
[text~=\\"2xl\\"] {
  font-size: 1.5rem;
  line-height: 2rem;
}
.opacity-10 {
  opacity: 0.1;
}
.p-5 {
  padding: 1.25rem;
}
[p~=\\"x-2\\"] {
  padding-left: 0.5rem;
  padding-right: 0.5rem;
}
[p~=\\"y-1\\"] {
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
}
[text~=\\"lime-400\\"] {
  --tw-text-opacity: 1;
  color: rgba(163, 230, 53, var(--tw-text-opacity));
}
[text~=\\"blue-400\\"] {
  --tw-text-opacity: 1;
  color: rgba(96, 165, 250, var(--tw-text-opacity));
}
.italic {
  font-style: italic;
}
@media (min-width: 768px) {
  [md~=\\"italic\\"] {
    font-style: italic;
  }
}"
`;
